File: src/core/project-move/paths.ts
// src/core/project-move/paths.ts — shared path normalization helpers.
//
// Phase 4a rev3 (Codex Q3 + Gemini): expandHome was previously inlined in
// src/tools/project.ts but the CLI and batch runner bypassed it. Centralize
// so every boundary uses the same rule.

import { homedir } from 'node:os';

/**
 * Expand a leading `~` or `~/...` to the user's home directory.
 * AI agents (and users) frequently generate `~/path`; Node's fs rejects it
 * with ENOENT. Normalize at the MCP / CLI / batch boundaries.
 *
 * Does NOT resolve relative paths — that's `path.resolve`'s job. Pure `~`
 * handling so the behavior is trivially explainable.
 */
export function expandHome(p: string): string {
  if (!p) return p;
  if (p === '~') return homedir();
  if (p.startsWith('~/')) return `${homedir()}/${p.slice(2)}`;
  return p;
}

File: macos/EngramCoreWrite/ProjectMove/Paths.swift
// macos/EngramCoreWrite/ProjectMove/Paths.swift
// Mirrors src/core/project-move/paths.ts (Node parity baseline).
//
// Pure `~`/`~/` expansion only — does not resolve relative paths. Centralized
// so MCP / CLI / batch boundaries apply the same rule.
import Foundation
import EngramCoreRead

public enum ProjectPath {
    /// Expand a leading `~` or `~/...` to the user's home directory. Empty
    /// strings and paths without a leading tilde pass through unchanged.
    public static func expandHome(
        _ path: String,
        homeDirectory: URL = FileManager.default.homeDirectoryForCurrentUser
    ) -> String {
        ProjectReviewPathSupport.expandHome(path, homeDirectory: homeDirectory)
    }
}
